File: src/main/java/com/previopls/entity/Lead.java
package com.previopls.entity;

import com.previopls.entity.enums.PrioridadeLead;
import com.previopls.entity.enums.StatusLead;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.Instant;
import java.util.UUID;

@Entity
@Table(name = "leads", indexes = {
        @Index(name = "ix_leads_cliente_id", columnList = "cliente_id"),
        @Index(name = "ix_leads_veiculo_id", columnList = "veiculo_id"),
        @Index(name = "ix_leads_prioridade", columnList = "prioridade"),
        @Index(name = "ix_leads_status", columnList = "status"),
        @Index(name = "ix_leads_prioridade_status", columnList = "prioridade, status")
})
@Getter
@Setter
@NoArgsConstructor
public class Lead {

    @Id
    @GeneratedValue
    @Column(columnDefinition = "uuid")
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "cliente_id", nullable = false)
    private Cliente cliente;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "veiculo_id", nullable = false)
    private Veiculo veiculo;

    @Column(name = "score_risco", nullable = false)
    private Double scoreRisco;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private PrioridadeLead prioridade;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private StatusLead status = StatusLead.ABERTO;

    @Column(name = "script_oferta", columnDefinition = "text")
    private String scriptOferta;

    @Column(columnDefinition = "text")
    private String observacao;

    @Column(name = "criado_em", nullable = false, updatable = false)
    private Instant criadoEm = Instant.now();

    @Column(name = "atualizado_em", nullable = false)
    private Instant atualizadoEm = Instant.now();

    @PreUpdate
    void preUpdate() {
        this.atualizadoEm = Instant.now();
    }
}

File: services/gateway/app/models/lead.py
import enum
import uuid
from datetime import datetime

from sqlalchemy import Column, DateTime, Enum, Float, ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.db.session import Base


class PrioridadeLead(str, enum.Enum):
    CRITICA = "critica"
    ALTA = "alta"
    MEDIA = "media"
    BAIXA = "baixa"


class StatusLead(str, enum.Enum):
    ABERTO = "aberto"
    AGENDADO = "agendado"
    RECUSADO = "recusado"
    SEM_CONTATO = "sem-contato"


class Lead(Base):
    __tablename__ = "leads"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    cliente_id = Column(UUID(as_uuid=True), ForeignKey("clientes.id", ondelete="CASCADE"),
                        nullable=False, index=True)
    veiculo_id = Column(UUID(as_uuid=True), ForeignKey("veiculos.id", ondelete="CASCADE"),
                        nullable=False, index=True)

    score_risco = Column(Float, nullable=False)
    prioridade = Column(
        Enum(PrioridadeLead, name="prioridade_lead", values_callable=lambda x: [e.value for e in x]),
        nullable=False, index=True,
    )
    status = Column(
        Enum(StatusLead, name="status_lead", values_callable=lambda x: [e.value for e in x]),
        nullable=False, default=StatusLead.ABERTO, index=True,
    )
    script_oferta = Column(Text, nullable=True)
    observacao = Column(Text, nullable=True)
    criado_em = Column(DateTime(timezone=True), default=datetime.utcnow, nullable=False)
    atualizado_em = Column(DateTime(timezone=True), default=datetime.utcnow,
                           onupdate=datetime.utcnow, nullable=False)

    cliente = relationship("Cliente", back_populates="leads")
    veiculo = relationship("Veiculo", back_populates="leads")
